### PR TITLE
BUG: GH11206 where pd.isnull did not consider numpy NaT null

### LINKED ISSUE #11206: BUG: accept np.datetime64('NaT') in an object dtyped as a null
from [SO](http://stackoverflow.com/questions/32863674/python-pandas-isnull-does-not-work-on-nat-in-object-dtype)

this should work

```
Out[3]: array([11, 22, 33, nan, numpy.datetime64('NaT')], dtype=object)

In [4]: pd.isnull(ser)
Out[4]: 
0    False
1    False
2    False
3     True
4    False
Name: my_series, dtype: bool
```

This could be updated: https://github.com/pydata/pandas/blob/master/pandas/tslib.pyx#L711, then ``lib.isnullobj`` then could use for null detection

### PR DESCRIPTION
**Pull request by kawochen**
closes #11206

### CODE REVIEW

**Comment by jreback** [comment]
lgtm.

can you run a perf check on the null benchmarks. just confirm nothing has substantially changed.

**Comment by kawochen** [comment]
Updated. Timing for `frame_methods.frame_isnull.time_frame_isnull` (where all data is non-`nan` `float`s) increased by less than 5%.
Added a benchmark where lots of the data is various types of null.
```
BENCHMARK                                    BEFORE      AFTER     FACTOR
...ame_isnull_has_null.time_frame_isnull    38.18ms   138.88ms   3.63779915x
```

I also changed the code slightly so that the logic is further short-circuited short-circuited.

**Comment by jreback** [comment]
the benchmark u are showing is 3x increase?

**Comment by kawochen** [comment]
Yes.That's the with lots of nulls. I will dig a little to see if that can be avoided.

**Comment by jreback** [comment]
these routines are hit *everywhere*, so i am for special casing certain things. IOW, that is why ``isnullobj`` exists. But it may make sense to have 2 versions of ``_checknull_with_nat``, because they are called on different array kinds. E.g if you have an int array then it obviously doesn't need to check for ``None,NaN,NaT`` at all. most of these cases are already handled, but IIRC, this is hotspotted. So just want to try to nail that down.

**Comment by kawochen** [comment]
Updated.

File: pandas/lib.pyx [hunk 1]
```diff
@@ -321,11 +320,11 @@ def isnullobj(ndarray[object] arr):
...
 
     n = len(arr)
-    result = np.zeros(n, dtype=np.uint8)
+    result = np.empty(n, dtype=np.uint8)
     for i from 0 <= i < n:
         val = arr[i]
-        result[i] = val is NaT or _checknull(val)
-    return result.view(np.bool_)
+        result[i] = _check_all_nulls(val)
+    return result.view(dtype=np.bool_)
```
Comment (kawochen) [inline pandas/lib.pyx]: I forgot to remove this.  I will remove in the next update.

Comment (jreback) [inline pandas/lib.pyx]: what are you fixing here?

Comment (kawochen) [inline pandas/lib.pyx]: you mean the `_check_all_nulls`?  Only object arrays hit this path so I thought it'd be safe to have the most generic version of `checknull` here?

For the `dtype` thing, I forgot to remove it after some experiment, but will remove it in the next update.

File: asv_bench/benchmarks/frame_methods.py [hunk 2]
```diff
@@ -582,7 +582,7 @@ def time_frame_interpolate_some_good_infer(self):
         self.df.interpolate(downcast='infer')
 
 
-class frame_isnull(object):
+class frame_isnull_flaots_no_null(object):
```
Comment (jreback) [inline asv_bench/benchmarks/frame_methods.py]: ``floats`` typo

File: pandas/lib.pyx (hunk 1 above)
Comment (jreback) [inline pandas/lib.pyx]: yep that is fine

File: pandas/tslib.pyx [hunk 3]
```diff
@@ -707,12 +708,26 @@ NaT = NaTType()
...
-
 cdef inline bint _checknull_with_nat(object val):
     """ utility to check if a value is a nat or not """
     return val is None or (
         PyFloat_Check(val) and val != val) or val is NaT
 
+cdef inline bint _check_all_nulls(object val):
+    """ utility to check if a value is any type of null """
+    cdef bint res = 0
+    if PyFloat_Check(val):
```
Comment (jreback) [inline pandas/tslib.pyx]: shouldn't you just return at each stage? rather than setting a value then returning it at the end

**Comment by kawochen** [comment]
udpated

File: doc/source/whatsnew/v0.17.0.txt [hunk 4]
```diff
@@ -1146,7 +1146,7 @@ Bug Fixes
 - Bug in ``Categorical.__iter__`` may not returning correct ``datetime`` and ``Period`` (:issue:`10713`)
 - Bug in indexing with a ``PeriodIndex`` on an object with a ``PeriodIndex`` (:issue:`4125`)
 - Bug in ``read_csv`` with ``engine='c'``: EOF preceded by a comment, blank line, etc. was not handled correctly (:issue:`10728`, :issue:`10548`)
-
+- Bug in ``isnull`` where ``numpy.datetime64('NaT')`` in a ``numpy.array`` was not determined to be null(:issue:`11206`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: let's put this in 0.17.1

Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: pls move to 0.17.1

**Comment by kawochen** [comment]
ok. I'm traveling and will get to this next week.

**Comment by jreback** [comment]
@kawochen can you update (move whatsnew to 0.17.1)

**Comment by kawochen** [comment]
Will do (still on vacation)

**Comment by kawochen** [comment]
updated

**Comment by jreback** [comment]
can u rebase and run a perf check

**Comment by kawochen** [comment]
Saw some horrifying perf hit (30X).  I will look at the compiled code later.

**Comment by jreback** [comment]
ok gr8!

**Comment by kawochen** [comment]
so using `timeit` and this code
```
from pandas import *
import pandas as pd
import numpy as np

np.random.seed(1234)
sample = np.array([NaT, np.nan, None, np.datetime64('NaT'),
    np.timedelta64('NaT'), 0, 1, 2.0, '', 'abcd'])
data = np.random.choice(sample, (1000, 1000))
df = DataFrame(data)

%timeit isnull(df)
```
the perf hit is 4X.  But if I that put into an `asv` benchmark, it shows 24X. I seem to be having problems with `asv` lately -- not sure what's wrong.

**Comment by jreback** [comment]
are you using ``profile=True`` in the cython modules?

**Comment by pv** [comment]
The perf hit is different for numpy 1.10 vs 1.9

**Comment by jreback** [comment]
@kawochen pushing, but if you finish in the next few days, pls ping

**Comment by kawochen** [comment]
@pv thanks! :smile: 
@jreback 
branch
`
[ 32.58%] ··· Running ...s.frame_isnull_obj.time_frame_isnull    46.07ms
`
master
`
[ 82.58%] ··· Running ...s.frame_isnull_obj.time_frame_isnull    38.91ms
`

`

File: pandas/tslib.pyx [hunk 5]
```diff
@@ -707,12 +708,28 @@ NaT = NaTType()
 
 iNaT = util.get_nat()
 
-
```
Comment (jreback) [inline pandas/tslib.pyx]: maybe blow away ``_checknull_with_nat``(and replace with ``_check_all_nulls``  and see if any timings are affected?

**Comment by kawochen** [comment]
Replacing wouldn't work in all cases, because sometimes you don't want `timedelta64` to be accepted, e.g. `datetime_to_datetime64()`, so it would become some sort of manual inlining for `_checknull_with_nat` to deal with those cases (which might not be a bad thing, depending on how often this is needed).

But I see that `tslib.pyx` needs some cleaning, as code like `if val is np_NaT or val.view('i8') == iNaT` is everywhere (`np_NaT` is not a singleton, so the first half of the expression is almost certainly `False`, and the `val.view('i8')` is slower than `get_datetime64_val` (especially given the regression in numpy 1.10, but that will be fixed soon).

So maybe when others or I clean up this module, we can see how to organize or deal with all the lower level scalar check null functions

**Comment by jreback** [comment]
@kawochen thanks!

always nice PR's from you! keep it up!

**Pull request merged**